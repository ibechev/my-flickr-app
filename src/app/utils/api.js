/* eslint-disable no-console */
const API_KEY = "6aacdfc2ae7edfa681a9bf5c55efd555";
const IMAGES_PER_PAGE = 20;

const getImagesBasicInfo = async ({ page, tags }) => {
  const tagsString = tags.join("%2C");
  const url = `https://api.flickr.com/services/rest/?method=flickr.photos.search&api_key=${API_KEY}&tags=${tagsString}&per_page=${IMAGES_PER_PAGE}&page=${page}&format=json&nojsoncallback=1`;

  return await fetch(url)
    .then(response => response.json())
    .catch(error => console.warn(error));
};

const buildImageObject = async ({ farm, id, secret, server, title }) => {
  const urlInfo = `https://api.flickr.com/services/rest/?method=flickr.photos.getInfo&api_key=${API_KEY}&photo_id=${id}&secret=${secret}&format=json&nojsoncallback=1`;
  const thumbnail = `https://farm${farm}.staticflickr.com/${server}/${id}_${secret}_n.jpg`;
  let flickrUrl = "";

  const response = await fetch(urlInfo);
  const {
    stat,
    photo: {
      owner: { username: owner },
      tags: { tag: tags },
      dates: { taken },
      urls: { url: urlArray }
    }
  } = await response.json();

  urlArray.forEach(url => {
    if (url.type === "photopage") flickrUrl = url._content;
  });

  if (stat !== "ok") {
    throw "Could't fetch photo details";
  } else {
    return {
      title,
      owner,
      tags: tags.map(tag => tag._content),
      taken,
      flickrUrl,
      thumbnail
    };
  }
};

export const getImages = async params => {
  const basicData = await getImagesBasicInfo(params).then(result => result);
  const {
    photos: { pages, photo },
    stat
  } = basicData;
  let promisesArray;

  try {
    if (stat === "ok") {
      promisesArray = photo.map(image => {
        return buildImageObject(image).then(res => res);
      });

      return {
        images: await Promise.all(promisesArray),
        pages
      };
    } else throw "There was a problem with Flickr api";
  } catch (error) {
    if (DEVELOPMENT) throw new Error(error);
  }
};
