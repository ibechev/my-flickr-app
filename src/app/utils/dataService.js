export default class DataService {
  // static #API_KEY = "6aacdfc2ae7edfa681a9bf5c55efd555";
  // static #IMAGES_PER_PAGE = 20;
}
