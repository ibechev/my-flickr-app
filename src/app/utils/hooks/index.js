import { useState, useEffect } from "react";

export const useLoadingDots = () => {
  const [dots, setDots] = useState([null]);

  useEffect(() => {
    const intervalId = setInterval(() => addDot(), 1000);
    return function cleanup() {
      clearInterval(intervalId);
    };
  });

  function addDot() {
    if (dots.length === 5) {
      setDots([]);
    } else {
      setDots([...dots, null]);
    }
  }

  return dots;
};
