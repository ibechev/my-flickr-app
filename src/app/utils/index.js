export const queryStringFromTags = tagsList =>
  `?${tagsList.map(tag => tag).join("&")}`;

export const formatDate = date => {
  return date
    .split(" ")[0]
    .split("-")
    .reverse()
    .join("-");
};

export const formatTitle = title => {
  let cutIndex;
  if (
    title.endsWith(".jpg") ||
    title.endsWith(".png") ||
    title.endsWith(".gif")
  ) {
    cutIndex = 4;
  } else if (title.endsWith(".jpeg")) {
    cutIndex = 5;
  } else cutIndex = 0;

  return title.slice(0, title.length - cutIndex);
};

export const encodeURITags = tags => tags.map(tag => encodeURIComponent(tag));
