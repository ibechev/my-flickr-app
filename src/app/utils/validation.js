/**
 *
 * @param {string} inputValue
 */
export const validateInput = (inputValue, tags) => {
  let error = "";

  if (typeof inputValue !== "string" && DEVELOPMENT)
    throw `Function 'validateInput' expects argument 'inputValue' to be string, but received ${typeof inputValue}`;

  if (tags.includes(inputValue)) error = "This tag already exist";
  return error;
};

/**
 *
 * @param {string[]} tags
 */
export const validateSearch = tags => {
  let error = "";

  if (!Array.isArray(tags) && DEVELOPMENT) {
    throw `Function 'validateSearch expects argument 'tags' to be an array, but received ${typeof tags}`;
  }

  if (!tags.length) error = "At least one tag is required";
  return error;
};
