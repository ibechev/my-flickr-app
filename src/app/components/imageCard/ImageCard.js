import React from "react";
import PropTypes from "prop-types";
import { formatTitle, formatDate } from "../../../app/utils";

const ImageCard = ({ title, owner, tags, taken, flickrUrl, thumbnail }) => {
  return (
    <li className="image-card">
      <div className="header">
        <div className="header-info">
          <h6>{title ? formatTitle(title) : "Unnamed"}</h6>
          <p>{`by: ${owner}`}</p>
          <p>{`taken on: ${formatDate(taken)}`}</p>
        </div>

        <div className="flickr-link">
          <a href={flickrUrl} target="_blanc" title="See on Flickr">
            <i className="fab fa-flickr" />
          </a>
        </div>
      </div>

      <div className="image-wrapper">
        <img src={thumbnail} alt={title} />
      </div>

      <div className="footer">
        <div className="tag-label">
          <i className="fas fa-tags" />
          <h6>Tags</h6>
        </div>
        <ul className="tags">
          {tags.map((tag, i) => (
            <li key={i} className="listed-tag">
              {tag}
            </li>
          ))}
        </ul>
      </div>
    </li>
  );
};

ImageCard.propTypes = {
  title: PropTypes.string.isRequired,
  owner: PropTypes.string.isRequired,
  tags: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
  taken: PropTypes.string.isRequired,
  flickrUrl: PropTypes.string.isRequired,
  thumbnail: PropTypes.string.isRequired
};

export default ImageCard;
