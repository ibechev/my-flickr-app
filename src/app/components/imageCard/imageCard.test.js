import React from "react";
import renderer from "react-test-renderer";
import ImageCard from "./ImageCard";
import { propTypeErrors } from "../../../../__tests__/testUtils";
import { image } from "../../../../__tests__/testValues";
import { shallow } from "enzyme";
import { formatTitle, formatDate } from "../../utils";

describe("Expect <ImageCard />", () => {
  const defaultProps = { ...image };

  it("to render correctly", () => {
    const wrapper = renderer.create(<ImageCard {...defaultProps} />).toJSON();
    expect(wrapper).toMatchSnapshot();
  });

  it("to not throw error with correct prop types", () => {
    propTypeErrors(<ImageCard {...defaultProps} />, defaultProps);
  });

  it("to format the title with 'formatTitle()' utility method", () => {
    const wrapper = shallow(<ImageCard {...defaultProps} />);
    const formatedTitle = formatTitle(defaultProps.title);
    expect(
      wrapper
        .find(".header-info")
        .find("h6")
        .text()
    ).toEqual(formatedTitle);
  });

  it("to format the date with 'formatDate()' utility method", () => {
    const wrapper = shallow(<ImageCard {...defaultProps} />);
    const formatedDate = `taken on: ${formatDate(defaultProps.taken)}`;
    expect(
      wrapper
        .find(".header-info")
        .find("p")
        .at(1)
        .text()
    ).toEqual(formatedDate);
  });
});
