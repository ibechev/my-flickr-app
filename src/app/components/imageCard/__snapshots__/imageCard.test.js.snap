// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect <ImageCard /> to render correctly 1`] = `
<li
  className="image-card"
>
  <div
    className="header"
  >
    <div
      className="header-info"
    >
      <h6>
        Image title
      </h6>
      <p>
        by: owner name
      </p>
      <p>
        taken on: 07-10-2018
      </p>
    </div>
    <div
      className="flickr-link"
    >
      <a
        href="https://www.some-url.com"
        target="_blanc"
        title="See on Flickr"
      >
        <i
          className="fab fa-flickr"
        />
      </a>
    </div>
  </div>
  <div
    className="image-wrapper"
  >
    <img
      alt="Image title.jpeg"
      src="https://farm5.staticflickr.com/4850/45579906624_4dbf8572a6_n.jpg"
    />
  </div>
  <div
    className="footer"
  >
    <div
      className="tag-label"
    >
      <i
        className="fas fa-tags"
      />
      <h6>
        Tags
      </h6>
    </div>
    <ul
      className="tags"
    >
      <li
        className="listed-tag"
      >
        tag1
      </li>
      <li
        className="listed-tag"
      >
        tag2
      </li>
      <li
        className="listed-tag"
      >
        tag3
      </li>
    </ul>
  </div>
</li>
`;
